### PR TITLE
Rpc tests

### PR DESCRIPTION
**Pull request by cchudant**
# Pull Request type

- Testing

## Does this introduce a breaking change?

No

## Other information

I will add test helpers along the way.

### CODE REVIEW

**Comment by cchudant** [comment]
done:
- block_hash_and_number
- all the get block rpcs
- get_transaction_{by_block_id_and_index,by_hash,receipt,status}
- get_state_update
- get_class_hash_at / get_nonce / get_storage_at 

next:
- get_class / get_class_at
- syncing
- get_events
- call / estimate_fee / estimate_message_fee and the trace transaction rpcs (execution rpcs)

I also migrated node and rpcs to using the new ChainConfig from #198

ChainConfig will get its own crate soon

File: crates/client/db/src/lib.rs [hunk 1]
```diff
@@ -66,33 +61,15 @@ pub(crate) async fn open_rocksdb(
 
     opts.set_env(&env);
 
-    let backup_hande = if let Some(backup_dir) = backup_dir {
-        let (restored_cb_sender, restored_cb_recv) = oneshot::channel();
-
-        let (sender, receiver) = mpsc::channel(1);
```
Comment (jbcaron) [inline crates/client/db/src/lib.rs:72]: what is the mpsc channel for?

File: crates/client/rpc/src/methods/read/get_block_with_receipts.rs [hunk 2]
```diff
@@ -63,3 +64,95 @@ pub fn get_block_with_receipts(
...
+    use starknet_core::types::{BlockTag, Felt, L1DataAvailabilityMode, ResourcePrice};
+
+    #[rstest]
+    fn test_get_block_with_receipts() {
+        let _ = env_logger::builder().is_test(true).try_init();
+        let (backend, rpc) = open_testing();
+        let SampleChain1 { block_hashes, expected_txs, expected_receipts, .. } = make_sample_chain_1(&backend);
+
+        // Block 0
+        let res = MaybePendingBlockWithReceipts::Block(BlockWithReceipts {
```
Comment (jbcaron) [inline crates/client/rpc/src/methods/read/get_block_with_receipts.rs:84]: #214  add default traits on primitives (`..Default::default()`)

File: crates/node/src/cli/mod.rs [hunk 3]
```diff
@@ -66,14 +73,43 @@ impl RunCmd {
         }
         self.name.as_ref().unwrap()
     }
+}
+
+/// Starknet network types.
+#[derive(Debug, Clone, Copy, clap::ValueEnum, PartialEq)]
+pub enum NetworkType {
```
Comment (jbcaron) [inline crates/node/src/cli/mod.rs:80]: you will need to add a custom network with a user-defined chainid and feeder url to be consistent with the ChainConfig type

**Review by jbcaron — APPROVED** [review]
- add error testing on `get_BLOCK...` methods
- add test for `get_storage`, uninitialized key on existing contract should return Felt::ZERO

File: crates/client/db/src/lib.rs (hunk 1 above)
Comment (cchudant) [reply crates/client/db/src/lib.rs:72]: backup restore task, i moved elsewhere it so that DeoxysBackend::open_for_testing is not async for convenience

when backups are enabled, a thread is spawned that owns the rocksdb BackupEngine (it is not thread safe) and it receives backup requests using a mpsc channel
There is also another oneshot channel involved: when restoring the db at startup, we want to wait for the backupengine to finish restoration before returning from `open()`

I will add a comment about that

File: crates/node/src/cli/mod.rs (hunk 3 above)
Comment (cchudant) [reply crates/node/src/cli/mod.rs:80]: yepyep coming in another pr

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 ## Next release
 
+- fix: pending contract storage not stored properly
```
Comment (apoorvsadana) [inline CHANGELOG.md:6]: I guess we can add more things here, wdyt?

File: crates/client/db/src/contract_db.rs [hunk 5]
```diff
@@ -46,6 +48,8 @@ impl DeoxysBackend {
                 // Get pending or fallback to latest block_n
                 let col = self.db.get_column(pending_col);
                 // todo: smallint here to avoid alloc
+
+                // NB: pending has keys in bincode, not bytes
```
Comment (apoorvsadana) [inline crates/client/db/src/contract_db.rs]: what's NB 😅?

File: crates/client/db/src/lib.rs [hunk 6]
```diff
@@ -66,33 +61,15 @@ pub(crate) async fn open_rocksdb(
...
-        log::debug!("blocking on db restoration");
-        restored_cb_recv.await.context("Restoring database")?;
-        log::debug!("done blocking on db restoration");
-
-        Some(sender)
-    } else {
-        None
-    };
-
+    #[cfg(not(feature = "testing"))]
```
Comment (apoorvsadana) [inline crates/client/db/src/lib.rs]: is this needed?

File: crates/client/db/src/lib.rs [hunk 7]
```diff
@@ -355,6 +334,7 @@ struct BackupRequest {
 
 impl Drop for DeoxysBackend {
     fn drop(&mut self) {
+        #[cfg(not(feature = "testing"))]
```
Comment (apoorvsadana) [inline crates/client/db/src/lib.rs]: is this needed?

File: crates/client/rpc/src/methods/read/block_hash_and_number.rs [hunk 8]
```diff
@@ -20,3 +20,56 @@ pub fn block_hash_and_number(starknet: &Starknet) -> StarknetRpcResult<BlockHash
...
+mod tests {
+    use super::*;
+    use crate::test_utils::open_testing;
+    use dp_block::{DeoxysBlockInfo, DeoxysBlockInner, DeoxysMaybePendingBlock, DeoxysMaybePendingBlockInfo, Header};
+    use dp_state_update::StateDiff;
+    use rstest::rstest;
+    use starknet_core::types::Felt;
+
+    #[rstest]
+    fn test_block_hash_and_number() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/block_hash_and_number.rs]: Should we also check that if a Pending block is present then the block returned by `block_hash_and_number` is the non pending one?

File: crates/client/rpc/src/methods/read/get_block_transaction_count.rs [hunk 9]
```diff
@@ -28,3 +28,30 @@ pub fn get_block_transaction_count(starknet: &Starknet, block_id: BlockId) -> St
...
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_1, open_testing, SampleChain1};
+    use rstest::rstest;
+    use starknet_core::types::BlockTag;
+
+    #[rstest]
+    fn test_get_block_transaction_count() {
+        let (backend, rpc) = open_testing();
+        let SampleChain1 { block_hashes, .. } = make_sample_chain_1(&backend);
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_block_transaction_count.rs]: let's use more descriptive names instead of 1/2?

File: crates/client/rpc/src/methods/read/get_block_with_receipts.rs [hunk 10]
```diff
@@ -63,3 +64,95 @@ pub fn get_block_with_receipts(
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_1, open_testing, SampleChain1};
+    use rstest::rstest;
+    use starknet_core::types::{BlockTag, Felt, L1DataAvailabilityMode, ResourcePrice};
+
+    #[rstest]
+    fn test_get_block_with_receipts() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_block_with_receipts.rs]: so from what I understand, we store the data inside the DB with `make_sample_chain_1` and then we fetch that data with `get_block_with_receipts`. this seems to test the receipt functionality but in my the view, the ideal RPC test case would

- (as a sequencer), execute a txn using the RPC method and then check if the receipt was built correctly
- (as a full node), sync from the gateway a txn and check if the receipt was built correctly (or do we get receipts from the gateway as well in the full node, I am not sure)

File: crates/client/rpc/src/methods/read/get_block_with_tx_hashes.rs [hunk 11]
```diff
@@ -63,3 +63,86 @@ pub fn get_block_with_tx_hashes(
...
+    use super::*;
+    use crate::test_utils::{make_sample_chain_1, open_testing, SampleChain1};
+    use rstest::rstest;
+    use starknet_core::types::{BlockTag, Felt, L1DataAvailabilityMode, ResourcePrice};
+
+    #[rstest]
+    fn test_get_block_with_tx_hashes() {
+        let _ = env_logger::builder().is_test(true).try_init();
+        let (backend, rpc) = open_testing();
+        let SampleChain1 { block_hashes, tx_hashes, .. } = make_sample_chain_1(&backend);
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_block_with_tx_hashes.rs]: same as above. currently it looks like we're checking db storage and not the RPC logic itself or maybe I am missing something

File: crates/client/rpc/src/methods/read/get_block_with_txs.rs [hunk 12]
```diff
@@ -67,3 +67,86 @@ pub fn get_block_with_txs(starknet: &Starknet, block_id: BlockId) -> RpcResult<M
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_1, open_testing, SampleChain1};
+    use rstest::rstest;
+    use starknet_core::types::{BlockTag, Felt, L1DataAvailabilityMode, ResourcePrice};
+
+    #[rstest]
+    fn test_get_block_with_txs() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_block_with_txs.rs]: same as above

File: crates/client/rpc/src/methods/read/get_class_hash_at.rs [hunk 13]
```diff
@@ -22,7 +22,46 @@ pub fn get_class_hash_at(starknet: &Starknet, block_id: BlockId, contract_addres
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_2, open_testing, SampleChain2};
+    use rstest::rstest;
+    use starknet_core::types::BlockTag;
+
+    #[rstest]
+    fn test_get_class_hash_at() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_class_hash_at.rs]: same as above

File: crates/client/rpc/src/methods/read/get_state_update.rs [hunk 14]
```diff
@@ -80,3 +80,55 @@ pub fn get_state_update(starknet: &Starknet, block_id: BlockId) -> StarknetRpcRe
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_2, open_testing, SampleChain2};
+    use rstest::rstest;
+
+    #[rstest]
+    fn test_get_state_update() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_state_update.rs]: same as above

File: crates/client/rpc/src/methods/read/get_storage_at.rs [hunk 15]
```diff
@@ -55,3 +55,74 @@ pub fn get_storage_at(
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{make_sample_chain_2, open_testing, SampleChain2};
+    use rstest::rstest;
+    use starknet_core::types::BlockTag;
+
+    #[rstest]
+    fn test_get_storage_at() {
```
Comment (apoorvsadana) [inline crates/client/rpc/src/methods/read/get_storage_at.rs]: same as above

**Review by apoorvsadana — COMMENTED** [review]
So I have some questions on how are we looking at RPC test cases. In my view, the test cases can be divided into two categories

## Sequencer RPC tests

Here, I think we should send the transaction via one RPC and then check if we got the expected result. For example, if we want to check transaction hash caclculation, then, we should first do a dummy invoke transaction and check if we got the expected transaction hash. Or, for example, if we want to check transaction count, we should send 5 txs, manually seal a block (like in Substrate) and then call the transaction count method and see if it worked correctly. I might be biased on the approach because this is how we did it in Madara previously. But curious to understand more about your approach here.

## Full node RPC tests

Here, it would be similar to above but instead of sending transaction we would ideally mock the feeder gateway responses. I believe @cchudant said something like this on Slack.

Also, one suggestion would be to maybe split RPC tests into different PRs :)

File: crates/client/rpc/src/methods/read/get_storage_at.rs [hunk 16]
```diff
@@ -55,3 +55,74 @@ pub fn get_storage_at(
...
+                        assert_eq!(
+                            get_storage_at(&rpc, contracts[contract_i], keys[key_i], block_n),
+                            Err(StarknetRpcApiError::ContractNotFound),
+                            "get storage at blockid {block_n:?}, contract #{contract_i}, key #{key_i} should not found"
+                        );
+                    }
+                }
+            }
+        };
+
```
Comment (Mohiiit) [inline crates/client/rpc/src/methods/read/get_storage_at.rs:98]: shall we be testing about `BLOCK_NOT_FOUND` as well

File: crates/client/rpc/src/methods/read/get_transaction_by_block_id_and_index.rs [hunk 17]
```diff
@@ -34,3 +34,54 @@ pub fn get_transaction_by_block_id_and_index(
...
+    use starknet_core::types::BlockTag;
+
+    #[rstest]
+    fn test_get_transaction_by_block_id_and_index() {
+        let _ = env_logger::builder().is_test(true).try_init();
+        let (backend, rpc) = open_testing();
+        let SampleChain1 { block_hashes, expected_txs, .. } = make_sample_chain_1(&backend);
+
+        // Block 0
+        assert_eq!(get_transaction_by_block_id_and_index(&rpc, BlockId::Number(0), 0).unwrap(), expected_txs[0]);
```
Comment (Mohiiit) [inline crates/client/rpc/src/methods/read/get_transaction_by_block_id_and_index.rs:52]: we can also add tests for `BLOCK_NOT_FOUND` and `INVALID_TXN_INDEX` as well

File: crates/client/rpc/src/methods/read/get_transaction_receipt.rs [hunk 18]
```diff
@@ -65,3 +65,44 @@ pub async fn get_transaction_receipt(
...
+        let res = TransactionReceiptWithBlockInfo {
+            receipt: expected_receipts[2].clone(),
+            block: ReceiptBlock::Block { block_hash: block_hashes[2], block_number: 2 },
+        };
+        assert_eq!(get_transaction_receipt(&rpc, tx_hashes[2]).unwrap(), res);
+
+        // Pending
+        let res =
+            TransactionReceiptWithBlockInfo { receipt: expected_receipts[3].clone(), block: ReceiptBlock::Pending };
+        assert_eq!(get_transaction_receipt(&rpc, tx_hashes[3]).unwrap(), res);
```
Comment (Mohiiit) [inline crates/client/rpc/src/methods/read/get_transaction_receipt.rs:105]: tests for `TxnHashNotFound` not present

File: crates/node/src/cli/mod.rs (hunk 3 above)
Comment (Mohiiit) [reply crates/node/src/cli/mod.rs:80]: was thinking about it as well, given for app chains it would be much more simpler to have feeder url in the params

File: crates/client/db/src/contract_db.rs (hunk 5 above)
Comment (cchudant) [reply crates/client/db/src/contract_db.rs]: nota bene, which is like `Note:` here or `Pay attention:`
should I stop putting NBs all over the codebase? I feel like this is a common enough abbreviation but i'm french and biased

File: crates/client/db/src/lib.rs (hunk 6 above)
Comment (cchudant) [reply crates/client/db/src/lib.rs]: no :)
i just wanted to have cleaner logs in tests but I can remove it

File: crates/client/db/src/lib.rs (hunk 7 above)
Comment (cchudant) [reply crates/client/db/src/lib.rs]: same, i'll probably remove it

File: crates/client/rpc/src/methods/read/block_hash_and_number.rs (hunk 8 above)
Comment (cchudant) [reply crates/client/rpc/src/methods/read/block_hash_and_number.rs]: sure why not

File: crates/client/rpc/src/methods/read/get_transaction_receipt.rs (hunk 18 above)
Comment (cchudant) [reply crates/client/rpc/src/methods/read/get_transaction_receipt.rs:105]: done

File: crates/client/rpc/src/methods/read/get_transaction_by_block_id_and_index.rs (hunk 17 above)
Comment (cchudant) [reply crates/client/rpc/src/methods/read/get_transaction_by_block_id_and_index.rs:52]: done

File: CHANGELOG.md (hunk 4 above)
Comment (cchudant) [reply CHANGELOG.md:6]: done

**Review by antiyro — APPROVED** [review]
